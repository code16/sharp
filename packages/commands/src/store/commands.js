

export const SET_COMMANDS = 'SET_COMMANDS';

export default {
    namespaced: true,

    state: {
        commands: null,
    },
    mutations: {
        [SET_COMMANDS](state, commands) {
            state.commands = commands;
        }
    },
    getters: {
        forType(state) {
            return type => state.commands ? state.commands[type] : null;
        }
    },
    actions: {
        update({ commit }, { commands }) {
            commit(SET_COMMANDS, commands);
        }
    },
}