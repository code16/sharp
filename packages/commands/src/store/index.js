
export { default as commands } from './commands';