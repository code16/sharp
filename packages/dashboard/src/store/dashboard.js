import {
    getDashboard,
    postDashboardCommand,
    getDashboardCommandFormData,
} from "../api";
import { filtersModule as filters } from 'sharp-filters';
import { commandsModule as commands } from 'sharp-commands';

export const UPDATE = 'UPDATE';
export const SET_DASHBOARD_KEY = 'SET_DASHBOARD_KEY';

export default {
    namespaced: true,
    modules: {
        filters,
        commands,
    },
    state: {
        dashboardKey: null,
        data: null,
        widgets: null,
        config: null,
        layout: null,
    },
    mutations: {
        [UPDATE](state, { data, layout, widgets, config }) {
            state.data = data;
            state.widgets = widgets;
            state.layout = layout;
            state.config = config;
        },
        [SET_DASHBOARD_KEY](state, dashboardKey) {
            state.dashboardKey = dashboardKey;
        }
    },
    actions: {
        update({ commit, dispatch }, { data, widgets, layout, config, filtersValues }) {
            commit(UPDATE, {
                data,
                widgets,
                layout,
                config,
            });
            return Promise.all([
                dispatch('filters/update', {
                    filters: config.filters,
                    values: filtersValues
                }),
                dispatch('commands/update', {
                    commands: config.commands
                })
            ]);
        },
        async get({ state, dispatch, getters }, { filtersValues }) {
            const data = await getDashboard({
                dashboardKey: state.dashboardKey,
                filters: getters['filters/getQueryParams'](filtersValues)
            });
            await dispatch('update', {
                data: data.data,
                widgets: data.widgets,
                layout: data.layout,
                config: data.config,
                filtersValues,
            });
        },
        postCommand({ state }, { command, data, query }) {
            return postDashboardCommand({
                dashboardKey: state.dashboardKey,
                commandKey: command.key,
                data,
                query,
            });
        },
        getCommandFormData({ state }, { command, query }) {
            return getDashboardCommandFormData({
                dashboardKey: state.dashboardKey,
                commandKey: command.key,
                query,
            });
        },
        setDashboardKey({ commit }, dashboardKey) {
            commit(SET_DASHBOARD_KEY, dashboardKey);
        }
    }
}