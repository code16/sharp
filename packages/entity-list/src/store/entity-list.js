import { filtersModule as filters } from 'sharp-filters';
import { postEntityListReorder } from "../api";


export const SET_ENTITY_KEY = 'SET_ENTITY_KEY';
export const SET_QUERY = 'SET_QUERY';

export default {
    namespaced: true,
    modules: {
        filters,
    },

    state() {
        return {
            entityKey: null,
            query: {},
        }
    },

    mutations: {
        [SET_ENTITY_KEY](state, entityKey) {
            state.entityKey = entityKey;
        },
        [SET_QUERY](state, query) {
            state.query = query;
        }
    },

    getters: {
        query(state) {
            return state.query;
        },
    },

    actions: {
        update({ dispatch }, { data, layout, config, filtersValues }) {
            return Promise.all([
                dispatch('filters/update', {
                    filters: config.filters,
                    values: filtersValues
                }),
            ]);
        },
        reorder({ state }, { instances }) {
            return postEntityListReorder({
                entityKey: state.entityKey,
                instances,
            });
        },
        setEntityKey({ commit }, entityKey) {
            commit(SET_ENTITY_KEY, entityKey);
        },
        setQuery({ commit }, query) {
            commit(SET_QUERY, query);
        }
    }
}