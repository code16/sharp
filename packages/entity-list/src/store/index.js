
export { default as entityList } from './entity-list';