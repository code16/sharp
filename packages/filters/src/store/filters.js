import Vue from 'vue';
import { parseRange, serializeRange } from "sharp";
import { getFiltersQueryParams, getFiltersValuesFromQuery, filterQueryKey } from '../util/query';

export const SET_FILTERS = 'SET_FILTERS';
export const SET_FILTER_VALUE = 'SET_FILTER_VALUE';

export default {
    namespaced: true,

    state: ()=>({
        filters: null,
        values: {},
    }),

    mutations: {
        [SET_FILTERS](state, filters) {
            state.filters = filters;
        },
        [SET_FILTER_VALUE](state, { key, value }) {
            Vue.set(state.values, key, value);
        },
    },

    getters: {
        value(state) {
            return key => state.values[key];
        },
        filters(state) {
            return state.filters ||Â []
        },
        values(state) {
            return state.values;
        },
        filter(state) {
            return key => (state.filters || []).find(filter => filter.key === key);
        },

        defaultValue() {
            return filter => (filter||{}).default;
        },
        isDateRange() {
            return filter => (filter||{}).type === 'daterange';
        },

        filterQueryKey() {
            return key => filterQueryKey(key);
        },
        getQueryParams(state, getters) {
            return values => {
                return getFiltersQueryParams(values, (value, key) =>
                    getters.serializeValue({ filter:getters.filter(key), value })
                );
            }
        },
        getValuesFromQuery() {
            return query => getFiltersValuesFromQuery(query);
        },
        resolveFilterValue(state, getters) {
            return ({ filter, value }) => {
                if(value == null) {
                    return getters.defaultValue(filter);
                }
                if(filter.multiple && !Array.isArray(value)) {
                    return [value];
                }
                if(getters.isDateRange(filter)) {
                    return parseRange(value);
                }
                return value;
            }
        },
        serializeValue(state, getters) {
            return ({ filter, value }) => {
                if(getters.isDateRange(filter)) {
                    return serializeRange(value);
                }
                return value;
            };
        },
        nextValues(state) {
            return ({ filter, value }) => {
                let base = state.values;
                if(filter.master) {
                    base = Object.keys(state.values).reduce((res, key) => ({ ...res, [key]:null }), {});
                }
                return { ...base, [filter.key]: value };
            };
        },
        nextQuery(state, getters) {
            return ({ filter, value }) => {
                return getters.getQueryParams(getters.nextValues({ filter, value }));
            }
        },
    },

    actions: {
        update({ commit, dispatch }, { filters, values }) {
            commit(SET_FILTERS, filters);

            return Promise.all(
                (filters || []).map(filter => {
                    dispatch('setFilterValue', {
                        filter,
                        value: (values || {})[filter.key]
                    })
                })
            );
        },
        setFilterValue({ commit, getters }, { filter, value }) {
            commit(SET_FILTER_VALUE, {
                key: filter.key,
                value: getters.resolveFilterValue({ filter, value })
            });
        },
    }
}