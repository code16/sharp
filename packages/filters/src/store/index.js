
export { default as filters } from './filters';