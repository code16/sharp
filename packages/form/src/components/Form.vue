<template>
    <div class="SharpForm" data-popover-boundary>
        <template v-if="ready">
            <slot
                name="action-bar"
                :props="actionBarProps"
                :listeners="actionBarListeners"
            />

            <template v-if="hasErrors">
                <div class="SharpNotification SharpNotification--error" role="alert">
                    <div class="SharpNotification__details">
                        <div class="SharpNotification__text-wrapper">
                            <p class="SharpNotification__title">{{ l('form.validation_error.title') }}</p>
                            <p class="SharpNotification__subtitle">{{ l('form.validation_error.description') }}</p>
                        </div>
                    </div>
                </div>
            </template>

            <TabbedLayout :layout="layout" ref="tabbedLayout">
                <template v-if="localized" v-slot:nav-prepend>
                    <LocaleSelect
                        :locale="currentLocale"
                        :locales="locales"
                        @change="handleLocaleChanged"
                    />
                </template>
                <template v-slot:default="{ tab }">
                    <Grid :rows="[tab.columns]" ref="columnsGrid" v-slot="{ itemLayout:column }">
                        <FieldsLayout
                            :layout="column.fields"
                            :visible="fieldVisible"
                            ref="fieldLayout"
                            v-slot="{ fieldLayout }"
                        >
                            <FieldDisplay
                                :field-key="fieldLayout.key"
                                :context-fields="transformedFields"
                                :context-data="data"
                                :field-layout="fieldLayout"
                                :locale="fieldLocale[fieldLayout.key]"
                                :read-only="isReadOnly"
                                :error-identifier="fieldLayout.key"
                                :config-identifier="fieldLayout.key"
                                :update-data="updateData"
                                :update-visibility="updateVisibility"
                                @locale-change="updateLocale"
                                ref="field"
                            />
                        </FieldsLayout>
                    </Grid>
                </template>
            </TabbedLayout>
        </template>
    </div>
</template>

<script>
    import {
        BASE_URL,
        getBackUrl,
        getListBackUrl,
        logError,
        showAlert,
    } from "sharp";

    import { TabbedLayout, Grid, Dropdown, DropdownItem, } from 'sharp-ui';
    import { Localization, DynamicView } from 'sharp/mixins';

    import FieldsLayout from './ui/FieldsLayout';
    import LocaleSelect from './ui/LocaleSelect';
    import localize from '../mixins/localize/form';

    import { getDependantFieldsResetData, transformFields } from "../util";


    const noop = ()=>{};

    export default {
        name:'SharpForm',
        extends: DynamicView,

        mixins: [ Localization, localize('fields')],

        components: {
            TabbedLayout,
            FieldsLayout,
            Grid,
            Dropdown,
            DropdownItem,
            LocaleSelect,
        },

        props:{
            entityKey: String,
            instanceId: String,

            /// Extras props for customization
            independant: {
                type:Boolean,
                default: false
            },
            ignoreAuthorizations: Boolean,
            props: Object
        },

        provide() {
            return {
                $form:this
            }
        },

        data() {
            return {
                ready: false,

                fields: null,
                authorizations: null,
                breadcrumb: null,
                config: null,

                errors:{},
                fieldLocale: {},
                locales: null,

                fieldVisible: {},
                uploadingFields: {},
                curFieldsetId:0,
            }
        },
        computed: {
            apiPath() {
                let path = `form/${this.entityKey}`;
                if(this.instanceId) path+=`/${this.instanceId}`;
                return path;
            },
            localized() {
                return Array.isArray(this.locales) && !!this.locales.length;
            },
            isSingle() {
                return this.config
                    ? this.config.isSingle
                    : false
            },
            isCreation() {
                return !this.isSingle && !this.instanceId;
            },
            isReadOnly() {
                if(this.ignoreAuthorizations) {
                    return false;
                }
                return this.isCreation
                    ? !this.authorizations.create
                    : !this.authorizations.update;
            },
            // don't show loading on creation
            synchronous() {
                return this.independant;
            },
            hasErrors() {
                return Object.keys(this.errors).some(errorKey => !this.errors[errorKey].cleared);
            },

            baseEntityKey() {
                return this.entityKey.split(':')[0];
            },

            downloadLinkBase() {
                return `/download/${this.entityKey}/${this.instanceId}`;
            },
            listUrl() {
                return `${BASE_URL}/list/${this.baseEntityKey}?restore-context=1`;
            },
            transformedFields() {
                return transformFields(this.fields, this.data);
            },

            currentLocale() {
                const locales = [...new Set(Object.values(this.fieldLocale))];
                return locales.length === 1 ? locales[0] : null;
            },
            isUploading() {
                return Object.values(this.uploadingFields)
                    .some(uploading => !!uploading);
            },
            actionBarProps() {
                return {
                    showSubmitButton: this.isCreation
                        ? !!this.authorizations.create
                        : !!this.authorizations.update,
                    showDeleteButton: !this.isCreation && !this.isSingle && !!this.authorizations.delete,
                    showBackButton: this.isReadOnly,
                    create: !!this.isCreation,
                    uploading: this.isUploading,
                }
            },
            actionBarListeners() {
                return {
                    'submit': this.handleSubmitClicked,
                    'delete': this.handleDeleteClicked,
                    'cancel': this.handleCancelClicked,
                }
            },
        },
        methods: {
            async updateData(key, value, { forced } = {}) {
                this.data = {
                    ...this.data,
                    ...(!forced ? getDependantFieldsResetData(this.fields, key,
                        field => this.fieldLocalizedValue(field.key, null),
                    ) : null),
                    [key]: this.fieldLocalizedValue(key, value),
                }
            },
            updateVisibility(key, visibility) {
                this.$set(this.fieldVisible, key, visibility);
            },
            updateLocale(key, locale) {
                this.$set(this.fieldLocale, key, locale);
            },
            handleLocaleChanged(locale) {
                this.fieldLocale = this.defaultFieldLocaleMap({ fields: this.fields, locales: this.locales }, locale);
            },
            mount({ fields, layout, data={}, authorizations={}, locales, breadcrumb, config }) {
                this.fields = fields;
                this.data = data;
                this.layout = this.patchLayout(layout);
                this.locales = locales;
                this.authorizations = authorizations;
                this.breadcrumb = breadcrumb;
                this.config = config;

                if(fields) {
                    this.fieldVisible = Object.keys(this.fields).reduce((res, fKey) => {
                        res[fKey] = true;
                        return res;
                    },{});
                    this.fieldLocale = this.defaultFieldLocaleMap({ fields, locales });
                }
                this.validate();
            },
            validate() {
                const localizedFields = Object.keys(this.fieldLocale);
                const alert = text => showAlert(text, {
                    title: 'Data error',
                    isError: true,
                });
                if(localizedFields.length > 0 && !this.locales.length) {
                    alert("Some fields are localized but the form hasn't any locales configured");
                }
            },
            handleError({response}) {
                if(response.status===422)
                    this.errors = response.data.errors || {};
            },

            patchLayout(layout) {
                if(!layout)return;
                let curFieldsetId = 0;
                let mapFields = layout => {
                    if(layout.legend)
                        layout.id = `${curFieldsetId++}#${layout.legend}`;
                    else if(layout.fields)
                        layout.fields.forEach(row => {
                            row.forEach(mapFields);
                        });
                };
                layout.tabs.forEach(t => t.columns.forEach(mapFields));
                return layout;
            },

            async init() {
                if(this.independant) {
                    this.mount(this.props);
                    this.ready = true;
                }
                else {
                    if(this.entityKey) {
                        await this.get();
                        this.ready = true;
                    }
                    else logError('no entity key provided');
                }
            },
            redirectToList() {
                location.href = getListBackUrl(this.breadcrumb);
            },
            redirectToParentPage() {
                location.href = getBackUrl(this.breadcrumb);
            },
            async submit({ postFn }={}) {
                if(this.isUploading) {
                    return;
                }
                try {
                    const response = postFn ? await postFn(this.data) : await this.post();
                    if(this.independant) {
                        this.$emit('submit', response);
                        return response;
                    }
                    else if(response.data.ok) {
                        this.$store.dispatch('setLoading', true);
                        this.redirectToParentPage();
                    }
                }
                catch(error) {
                    this.handleError(error);
                    return Promise.reject(error);
                }
            },
            handleSubmitClicked() {
                this.submit().catch(()=>{});
            },
            handleDeleteClicked() {
                this.axiosInstance.delete(this.apiPath)
                    .then(() => {
                        this.redirectToList();
                    });
            },
            handleCancelClicked() {
                this.redirectToParentPage();
            },

            // Used by VueClip as injection
            setUploading(fieldKey, uploading) {
                this.uploadingFields = {
                    ...this.uploadingFields,
                    [fieldKey]: uploading
                }
            },

            // Used by List field as injection
            hasUploadingFields(fieldKey) {
                return Object.keys(this.uploadingFields)
                    .some(uploadingField => uploadingField.startsWith(`${fieldKey}.`));
            },
        },
        created() {
            this.$on('error-cleared', errorId => {
                if(this.errors[errorId])
                    this.$set(this.errors[errorId],'cleared',true);
            });
        },
        mounted() {
            this.init();
        }
    }
</script>