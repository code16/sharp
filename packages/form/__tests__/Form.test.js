import Vue from 'vue';
import Vuex from 'vuex';
import axios from 'axios';
import Form from '../src/components/Form.vue';
import store from 'sharp/store';

import { wait, MockI18n, nextRequestFulfilled } from "@sharp/test-utils";
import moxios from 'moxios';
import { shallowMount, createLocalVue } from '@vue/test-utils';

jest.mock('sharp');


describe('sharp-form', ()=>{
    Vue.use(MockI18n);

    function createWrapper({ propsData } = {}) {
        const localVue = createLocalVue();
        localVue.use(Vuex);

        return shallowMount(Form, {
            propsData: {
                entityKey: 'spaceship',
                instanceId: null,
                ignoreAuthorizations: null,
                ...propsData,
            },
            provide: {
                axiosInstance: axios.create(),
            },
            store: new Vuex.Store(store),
            // language=Vue
            stubs: {
                Grid:
                    `<div id="MOCKED_GRID">
                        <slot :item-layout="rows[0][0]" />
                    </div>`,
                FieldsLayout:
                    `<div id="MOCKED_FIELDS_LAYOUT"> 
                        <slot v-if="layout[0][0]" :field-layout="layout[0][0]" />
                    </div>`,
                FieldDisplay: true,
                TabbedLayout:
                    `<div id="MOCKED_TABBED_LAYOUT">
                        <slot :tab="layout.tabs[0]" />
                    </div>`,
            },
            created() {
                moxios.install(this.axiosInstance);
                moxios.uninstall = moxios.uninstall.bind(moxios, this.axiosInstance);
            },
            localVue
        });
    }

    function createLayout(fields) {
        return {
            tabs : [
                {
                    columns: [
                        {
                            fields,
                        }
                    ]
                }
            ]
        }
    }

    function createForm({ key='title', type='text', ...props } = {}) {
        return {
            data: { [key]: null },
            fields: {
                [key]: {
                    key,
                    type,
                    ...props,
                }
            },
            layout: createLayout([[{ key }]]),
        }
    }


    const oldDelay = moxios.delay;
    moxios.delay = 10;

    afterAll(() => {
        moxios.delay = oldDelay;
    });
    // beforeEach(() => {
    //     moxios.install();
    // });
    afterEach(()=>{
        moxios.uninstall();
    });

    test('can mount sharp-form', async ()=>{
        const wrapper = createWrapper();

        await nextRequestFulfilled({
            status: 200,
            response: createForm(),
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount "independant" sharp-form', ()=>{
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: createForm()
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount sharp-form with alert', ()=>{
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: createForm()
            }
        });

        wrapper.setData({
            errors: {
                title: ['required']
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('localized', ()=>{
        const wrapper = createWrapper();
        expect(wrapper.vm.localized).toBe(false);
        wrapper.setData({
            locales: ['fr', 'en']
        });
        expect(wrapper.vm.localized).toBe(true);
    });

    test('detect when is creation', ()=>{
        const wrapper = createWrapper();

        expect(wrapper.vm.isCreation).toBe(true);

        wrapper.setProps({
            instanceId: '10',
        });

        expect(wrapper.vm.isCreation).toBe(false);

    });

    test('is read only', () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: {
                    ...createForm(),
                    authorizations: {
                        create: false,
                        update: false
                    }
                }
            }
        });
        expect(wrapper.vm.isReadOnly).toBe(true);

        wrapper.setData({
            authorizations: {
                create: true,
                update: false,
            }
        });
        expect(wrapper.vm.isReadOnly).toBe(false);

        wrapper.setProps({
            instanceId: '10',
        });
        expect(wrapper.vm.isReadOnly).toBe(true);

        wrapper.setData({
            authorizations: {
                create: true,
                update: true,
            }
        });
        expect(wrapper.vm.isReadOnly).toBe(false);

        wrapper.setProps({
            ignoreAuthorizations: true,
        });
        wrapper.setData({
            authorizations: {
                create: false,
                update: false,
            }
        });
        expect(wrapper.vm.isReadOnly).toBe(false);
    })

    test('synchronous', () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: createForm(),
            }
        });

        expect(wrapper.vm.synchronous).toBe(true);

        wrapper.setProps({ independant: false });

        expect(wrapper.vm.synchronous).toBe(false);
    });


    test('has errors', () => {
        const wrapper = createWrapper();
        expect(wrapper.vm.hasErrors).toBe(false);

        wrapper.setData({
            errors: {
                field: ['required']
            }
        });

        expect(wrapper.vm.hasErrors).toBe(true);

        wrapper.setData({
            errors: {
                field: { cleared:true },
            }
        });

        expect(wrapper.vm.hasErrors).toBe(false);
    });

    test('expose appropriate props to layout components', () => {
        const wrapper = createWrapper({
            propsData: {
                independant: true,
                props: createForm()
            }
        });

        let { tabbedLayout, columnsGrid, fieldLayout } = wrapper.vm.$refs;

        expect(tabbedLayout.$options.propsData).toEqual({
            layout: { tabs:[{ columns:[{ fields:[[{key:'title'}]] }] }] } // $form.layout
        });
        expect(columnsGrid.$options.propsData).toEqual({
            rows: [[{ fields:[[{key:'title'}]] }]] // [$form.layout.tabs[0].columns]
        });
        expect(fieldLayout.$options.propsData).toEqual({
            layout: [[{key:'title'}]], // $form.layout.tabs[0].columns[0].fields
            visible: { title: true } // $form.fieldVisible
        });
    });

    test('expose appropriate props to field', async () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: {
                    ...createForm({ localized: true }),
                    locales: ['fr', 'en']
                }
            }
        });
        await wrapper.vm.$nextTick();
        const field = wrapper.find({ ref:'field' }).vm;

        expect(field.$attrs).toMatchObject({
            'field-key': 'title',
            'field-layout': { key: 'title' },
            'update-data': wrapper.vm.updateData,
            'locale': 'fr',
            'error-identifier': 'title',
            'config-identifier': 'title',
            'update-visibility': wrapper.vm.updateVisibility,
            'context-data': {
                title: null
            },
            'context-fields': {
                title: {
                    type: 'text',
                },
            },
        });
    });

    test('update data', () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: createForm(),
            }
        });
        wrapper.vm.fieldLocalizedValue = jest.fn(()=>'fieldLocalizedValue');
        expect(wrapper.vm.data.title).toBe(null);

        wrapper.vm.updateData('title', 'text');
        expect(wrapper.vm.fieldLocalizedValue).toHaveBeenCalledWith('title', 'text');
        expect(wrapper.vm.data.title).toBe('fieldLocalizedValue');
    });


    test('update visibility', async () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: createForm()
            }
        });

        expect(wrapper.vm.fieldVisible.title).toBe(true);

        wrapper.vm.updateVisibility('title', false);

        expect(wrapper.vm.fieldVisible.title).toBe(false);
    });

    test('mount', () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: {
                    ...createForm({ localized:true }),
                    authorizations: {
                        create: true,
                        update: false
                    },
                    locales: ['en', 'fr']
                }
            }
        });

        expect(wrapper.vm).toMatchObject({
            ...createForm({ localized:true }),
            authorizations: {
                create: true,
                update: false
            },
            locales: ['en', 'fr'],
            fieldLocale: {
                title: 'en'
            }
        });

        wrapper.vm.patchLayout = ()=>{};
        wrapper.vm.ready = false;
        wrapper.vm.mount({ fields:{}, locales:null });

        expect(wrapper.vm.fieldLocale).toEqual({ title:undefined });
    });

    test('mount async', async () => {
        const wrapper = createWrapper();

        await nextRequestFulfilled({
            status: 200,
            response: {
                ...createForm(),
                authorizations: {
                    create: true,
                    update: false
                }
            }
        });

        expect(wrapper.vm).toMatchObject({
            ...createForm(),
            authorizations: {
                create: true,
                update: false
            }
        });
    });

    test('handle 422', async () => {
        const wrapper = createWrapper();

        wrapper.vm.submit().catch(() => {});

        await nextRequestFulfilled({
            status: 422,
            response: {
                errors: {
                    title: ['invalid']
                }
            }
        });

        expect(wrapper.vm.errors).toEqual({
            title: ['invalid']
        });
    });

    test('patch layout', async () => {
        const wrapper = createWrapper({
            propsData:{
                independant: true,
                props: {
                    ...createForm(),
                    layout: createLayout( [
                        [{ legend:'fieldset', fields: [[]] }],
                        [{ legend:'fieldset', fields: [[]] }]
                    ]),
                }
            }
        });

        let {
            tabs : [
                {
                    columns: [{ fields }]
                }
            ]
        } = wrapper.vm.layout;

        expect(typeof fields[0][0].id).toBe('string');
        expect(fields[0][0].id).not.toEqual(fields[1][0].id)
    });


    test('setup action bar correctly', async () => {
        const wrapper = createWrapper();

        await nextRequestFulfilled({
            status: 200,
            response: {
                ...createForm(),
                authorizations: {
                    create: false,
                    update: false,
                    delete: false,
                }
            }
        });

        expect(wrapper.vm.actionBarProps).toMatchObject({
            showSubmitButton: false,
            showDeleteButton: false,
            showBackButton: true,
            create: true,
        });


        wrapper.vm.authorizations.create = true;

        expect(wrapper.vm.actionBarProps).toMatchObject({
            showSubmitButton: true,
            showDeleteButton: false,
            showBackButton: false,
            create: true,
        });


        wrapper.setProps({
            instanceId: '10',
        });
        wrapper.vm.authorizations.delete = true;

        expect(wrapper.vm.actionBarProps).toMatchObject({
            showSubmitButton: false,
            showDeleteButton: true,
            showBackButton: true,
            create: false,
        });

        wrapper.vm.authorizations.update = true;

        expect(wrapper.vm.actionBarProps).toMatchObject({
            showSubmitButton: true,
            showDeleteButton: true,
            showBackButton: false,
            create: false,
        });
    });

    test('redirect to list', async () => {
        const wrapper = createWrapper();

        expect(wrapper.vm.listUrl).toEqual('/sharp/list/spaceship?restore-context=1');

        wrapper.vm.redirectToList = jest.fn();

        wrapper.vm.handleDeleteClicked();

        await nextRequestFulfilled({
            status: 200
        }, 0);

        expect(wrapper.vm.redirectToList).toHaveBeenCalledTimes(1);

    });

    test('submit', async () => {
        const wrapper = createWrapper({
            propsData: {
                independant:true,
                props: createForm(),
            },
        });

        wrapper.vm.post = jest.fn(()=>Promise.resolve({ data: { ok: true } }));

        wrapper.vm.uploadingFields = { field:true };
        wrapper.vm.submit();

        await wait(10);

        expect(wrapper.vm.post).not.toHaveBeenCalled();


        wrapper.vm.uploadingFields = {}
        wrapper.vm.submit();

        await wait(10);

        expect(wrapper.vm.post).toHaveBeenCalledTimes(1);
    });

    test('dependant submit', async () => {
        const wrapper = createWrapper();

        wrapper.vm.post = jest.fn(()=>Promise.resolve({ data: { ok: true } }));
        wrapper.vm.handleError = jest.fn();

        wrapper.vm.submit().catch(() => {});

        await wait(10);

        expect(wrapper.vm.post).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.post.mock.calls[0][0]).toBeUndefined();
        expect(wrapper.vm.post.mock.calls[0][1]).toBeUndefined();


        expect(wrapper.vm.handleError).not.toHaveBeenCalled();

        wrapper.vm.post = jest.fn(()=>Promise.reject({ error: true }));

        wrapper.vm.submit().catch(() => {});

        await wait(10);

        expect(wrapper.vm.handleError).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.handleError).toHaveBeenCalledWith({ error: true });
    });

    test('delete', async ()=>{
        // refer 'redirect to list' test
    });

    test('cancel', async ()=>{
        // refer 'redirect to list' test
    });

    test('has localize mixin with right fieldsProps', async () => {
        const wrapper = await createWrapper();
        expect(wrapper.vm.$options._localizedForm).toBe('fields');
    });
});