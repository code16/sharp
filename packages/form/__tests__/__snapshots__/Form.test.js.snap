// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sharp-form can mount "independant" sharp-form 1`] = `
<div data-popover-boundary="" class="SharpForm">
	<!---->
	<div id="MOCKED_TABBED_LAYOUT">
		<div id="MOCKED_GRID">
			<div id="MOCKED_FIELDS_LAYOUT">
				<fielddisplay-stub field-key="title" context-fields="[object Object]" context-data="[object Object]" field-layout="[object Object]" read-only="true" error-identifier="title" config-identifier="title" update-data="function () { [native code] }" update-visibility="function () { [native code] }"></fielddisplay-stub>
			</div>
		</div>
	</div>
</div>
`;

exports[`sharp-form can mount sharp-form 1`] = `
<div data-popover-boundary="" class="SharpForm">
	<!---->
	<div id="MOCKED_TABBED_LAYOUT">
		<div id="MOCKED_GRID">
			<div id="MOCKED_FIELDS_LAYOUT">
				<fielddisplay-stub field-key="title" context-fields="[object Object]" context-data="[object Object]" field-layout="[object Object]" read-only="true" error-identifier="title" config-identifier="title" update-data="function () { [native code] }" update-visibility="function () { [native code] }"></fielddisplay-stub>
			</div>
		</div>
	</div>
</div>
`;

exports[`sharp-form can mount sharp-form with alert 1`] = `
<div data-popover-boundary="" class="SharpForm">
	<div role="alert" class="SharpNotification SharpNotification--error">
		<div class="SharpNotification__details">
			<div class="SharpNotification__text-wrapper">
				<p class="SharpNotification__title">{{ form.validation_error.title }}</p>
				<p class="SharpNotification__subtitle">{{ form.validation_error.description }}</p>
			</div>
		</div>
	</div>
	<div id="MOCKED_TABBED_LAYOUT">
		<div id="MOCKED_GRID">
			<div id="MOCKED_FIELDS_LAYOUT">
				<fielddisplay-stub field-key="title" context-fields="[object Object]" context-data="[object Object]" field-layout="[object Object]" read-only="true" error-identifier="title" config-identifier="title" update-data="function () { [native code] }" update-visibility="function () { [native code] }"></fielddisplay-stub>
			</div>
		</div>
	</div>
</div>
`;
